bvm/ballerina-core/src/main/java/org/ballerinalang/util/metrics/MetricRegistry.java

```diff
@@ -55,80 +51,133 @@ public Counter counter(MetricId id) {
-    public Gauge gauge(MetricId id) {
-        return getOrCreate(id, Gauge.class, () -> metricProvider.newGauge(id));
+    public Gauge gauge(MetricId id, StatisticConfig... statisticConfigs) {
+        return getOrCreate(id, Gauge.class, () -> metricProvider.newGauge(id, statisticConfigs));
     }
 
     /**
-     * Use {@link CallbackGauge#builder(String, Object, ToDoubleFunction)}.
+     * Registers the gauge metrics instance.
+     *
+     * @param gauge The {@link Gauge} instacne.
+     */
+    public Gauge register(Gauge gauge) {
+        return register(gauge, Gauge.class);
+    }
+
+    /**
+     * Unregister the gauge metrics instance.
+     *
+     * @param gauge The {@link Gauge} instacne.
+     */
+    public void unregister(Gauge gauge) {
+        unregister(gauge, Gauge.class);
+    }
+
+    /**
+     * Use {@link PolledGauge#builder(String, Object, ToDoubleFunction)}.
      *
      * @param id            The {@link MetricId}.
      * @param obj           State object used to compute a value.
      * @param valueFunction Function that produces an instantaneous gauge value from the state object.
      * @param <T>           The type of the state object from which the gauge value is extracted.
-     * @return A existing or a new {@link CallbackGauge} metric.
+     * @return A existing or a new {@link PolledGauge} metric.
      */
-    public <T> CallbackGauge callbackGauge(MetricId id, T obj, ToDoubleFunction<T> valueFunction) {
-        return getOrCreate(id, CallbackGauge.class, () -> metricProvider.newCallbackGauge(id, obj, valueFunction));
+    public <T> PolledGauge polledGauge(MetricId id, T obj, ToDoubleFunction<T> valueFunction) {
+        return getOrCreate(id, PolledGauge.class, () -> metricProvider.newPolledGauge(id, obj, valueFunction));
     }
 
     /**
-     * Use {@link Summary#builder(String)}.
+     * Registers the polled gauge metrics instance.
      *
-     * @param id The {@link MetricId}.
-     * @return A existing or a new {@link Summary} metric.
+     * @param gauge The {@link PolledGauge} instacne.
      */
-    public Summary summary(MetricId id) {
-        return getOrCreate(id, Summary.class, () -> metricProvider.newSummary(id));
+    public PolledGauge register(PolledGauge gauge) {
+        return register(gauge, PolledGauge.class);
     }
 
     /**
-     * Use {@link Timer#builder(String)}.
+     * Unregisters the polled gauge metrics instance.
      *
-     * @param id The {@link MetricId}.
-     * @return A existing or a new {@link Timer} metric.
+     * @param gauge The {@link PolledGauge} instacne.
      */
-    public Timer timer(MetricId id) {
-        return getOrCreate(id, Timer.class, () -> metricProvider.newTimer(id));
+    public void unregister(PolledGauge gauge) {
+        unregister(gauge, PolledGauge.class);
     }
 
     private <M extends Metric> M getOrCreate(MetricId id, Class<M> metricClass, Supplier<M> metricSupplier) {
-        long stamp = stampedLock.tryOptimisticRead();
```

Comment:
I prefer to have this stamped lock. With the new changes `getOrCreate` method may create multiple metrics when there uses access this method concurrently. This may impact the performance negatively. I used `StampedLock` because it is much better under high concurrency. See http://isuru-perera.blogspot.com/2016/05/benchmarking-java-locks-with-counters.html

Reply:
Though the lock is obtained optimistically, we anyhow we need to read the metrics ConcurrentHashMap, therefore when we use concurrentHashMap I don't see the necessity of another locking mechanism.

Reply:
But now there is a new object created before checking from the map. I think it's better to avoid that.

Reply:
No the new object is created after checking the map. You can see below, the readMetric() method will check and return the element in the map, and only if it's null, we create a new element.

Reply:
It's okay to remove the additional locking mechanism since Metrics will be created only once anyway.